### PR TITLE
Deal feature

### PR DESCRIPTION
**Pull request by munkhjin0223**

### CODE REVIEW

File: src/modules/common/utils/listObjectUnFreeze.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const listObjectUnFreeze = source => {
```
Comment (batamar) [inline src/modules/common/utils/listObjectUnFreeze.js]: Remove this util

File: src/modules/deals/components/Stage.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+              >
+                <StageHeader {...provided.dragHandleProps}>
+                  <div>
+                    <h3>{stage.name}</h3>
+                    <span className="deals-count">
+                      {__('Deal')}: {deals.length}
+                    </span>
+                  </div>
+                  <ul>
+                    {Object.keys(amount).length > 0 ? (
```
Comment (batamar) [inline src/modules/deals/components/Stage.js]: write seperate function

File: src/modules/deals/components/Board.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      </BarItems>
+    );
+
+    const actionBar = (
+      <Wrapper.ActionBar left={actionBarLeft} background="transparent" />
+    );
+
+    const stageStates = {};
+
+    Object.keys(this.state).forEach(key => {
```
Comment (batamar) [inline src/modules/deals/components/Board.js]: use filter function

File: src/modules/deals/components/deal/Deal.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          <div>
+            <DealContainer
+              innerRef={provided.innerRef}
+              {...provided.draggableProps}
+              {...provided.dragHandleProps}
+            >
+              <DealHeader>
+                <h4>{deal.customer.firstName || deal.customer.email}</h4>
+                <span>{moment(deal.closeDate).format('YYYY-MM-DD')}</span>
+              </DealHeader>
```
Comment (batamar) [inline src/modules/deals/components/deal/Deal.js]: Add spaces above and below component

File: src/modules/deals/components/deal/DealForm.js [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+      company,
+      closeDate,
+      productsData,
+      note,
+      assignedUserIds
+    } = this.state;
+
+    if (!company) {
+      Alert.error('Choose a company');
+      return;
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: Just return Alert

File: src/modules/deals/components/deal/DealForm.js [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+    }
+
+    if (!customer) {
+      Alert.error('Choose a customer');
+      return;
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: Just return Alert

File: src/modules/deals/components/deal/DealForm.js [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+            title="Select company"
+            trigger={companyTrigger}
+          >
+            <CompanyAssociate
+              data={{ firstName: 'Deal', companies: company ? [company] : [] }}
+              save={this.onChangeCompany}
+              limit={1}
+            />
+          </ModalTrigger>
+          {company ? (
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: write seperate function

File: src/modules/deals/components/deal/DealForm.js [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+            <FormGroup>
+              <DealProducts>
+                <ul>
+                  <li>{company.name}</li>
+                </ul>
+              </DealProducts>
+            </FormGroup>
+          ) : null}
+          {company ? (
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: write seperate function

File: src/modules/deals/components/deal/DealForm.js [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+                  name: company.name,
+                  customers: customer ? [customer] : []
+                }}
+                companyId={company._id}
+                save={this.onChangeCustomer}
+                limit={1}
+              />
+            </ModalTrigger>
+          ) : null}
+          {customer ? (
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: write seperate function

File: src/modules/deals/components/deal/DealForm.js [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+            <FormGroup>
+              <DealProducts>
+                <ul>
+                  <li>{renderFullName(customer)}</li>
+                </ul>
+              </DealProducts>
+            </FormGroup>
+          ) : null}
+          {Object.keys(amount).length !== 0 ? (
```
Comment (batamar) [inline src/modules/deals/components/deal/DealForm.js]: write seperate function

File: src/modules/deals/components/deal/DealMoveForm.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    };
+  }
+
+  move() {
+    const { deal, boardId, pipelineId } = this.props;
+    const { stageId } = this.state;
+
+    if (!stageId) {
+      Alert.error('No stage');
+      return;
```
Comment (batamar) [inline src/modules/deals/components/deal/DealMoveForm.js]: Just run Alert

File: src/modules/deals/components/deal/DealProduct.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    if (length === 0) {
+      return null;
+    }
+
+    const otherProducts = () => {
+      if (this.state.show) {
+        return products.map(
+          (el, index) => (index > 0 ? <li key={el._id}>{el.name}</li> : null)
+        );
+      } else {
```
Comment (batamar) [inline src/modules/deals/components/deal/DealProduct.js]: else is redundant

File: src/modules/deals/components/deal/DealUser.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            <img
+              alt={el.details.fullName || el.email}
+              src={el.details.avatar || '/images/avatar-colored.svg'}
+            />
+          </Tip>
+        </li>
+      );
+    };
+
+    const otherUsers = () => {
```
Comment (batamar) [inline src/modules/deals/components/deal/DealUser.js]: write outside of render

File: src/modules/deals/components/deal/DealUser.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+      if (this.state.show) {
+        return users.map((el, index) => (index > 0 ? userItem(el) : null));
+      } else {
```
Comment (batamar) [inline src/modules/deals/components/deal/DealUser.js]: else is redundant

File: src/modules/deals/components/deal/DealUser.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+  render() {
+    const users = this.props.users;
+    const length = users.length;
+
+    if (length === 0) {
+      return null;
+    }
+
+    const userItem = el => {
```
Comment (batamar) [inline src/modules/deals/components/deal/DealUser.js]: write outside of render

File: src/modules/deals/components/product/ProductForm.js [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+          <Icon icon="plus" /> {__('Add Product / Service')}
+        </AddProduct>
+        <ProductFooter>
+          <FooterInfo>
+            <table>
+              <tbody>
+                <tr>
+                  <td>{__('Tax')}:</td>
+                  <td>
+                    {Object.keys(tax).map(t => (
```
Comment (batamar) [inline src/modules/deals/components/product/ProductForm.js]: write seperate function

File: src/modules/deals/components/product/ProductForm.js [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+                      <div key={t}>
+                        {tax[t].toLocaleString()} {t}
+                      </div>
+                    ))}
+                  </td>
+                </tr>
+                <tr>
+                  <td>{__('Discount')}:</td>
+                  <td>
+                    {Object.keys(discount).map(d => (
```
Comment (batamar) [inline src/modules/deals/components/product/ProductForm.js]: write separate function

File: src/modules/deals/components/product/ProductForm.js [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+                      <div key={d}>
+                        {discount[d].toLocaleString()} {d}
+                      </div>
+                    ))}
+                  </td>
+                </tr>
+                <tr>
+                  <td>{__('Total')}:</td>
+                  <td>
+                    {Object.keys(total).map(t => (
```
Comment (batamar) [inline src/modules/deals/components/product/ProductForm.js]: write separate function

File: src/modules/deals/containers/deal/DealMoveForm.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Spinner } from 'modules/common/components';
```
Comment (batamar) [inline src/modules/deals/containers/deal/DealMoveForm.js]: Always import third party packages above

File: src/modules/deals/containers/Board.js [hunk 20]
```diff
@@ -0,0 +1,143 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { withRouter } from 'react-router';
+import { router as routerUtils } from 'modules/common/utils';
+import { Spinner } from 'modules/common/components';
+import { compose, graphql } from 'react-apollo';
```
Comment (batamar) [inline src/modules/deals/containers/Board.js]: Import third party packages above

File: src/modules/deals/components/deal/Deal.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Draggable } from 'react-beautiful-dnd';
+import moment from 'moment';
+import { DealProduct, DealUser } from '../';
+import { QuickEdit } from '../';
```
Comment (batamar) [inline src/modules/deals/components/deal/Deal.js]: Import in above line

File: src/modules/deals/components/Board.js [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+        <Dropdown id="dropdown-board">
+          <DropdownToggle bsRole="toggle">
+            <Button btnStyle="simple" size="small">
+              {currentBoard.name} <Icon icon="ios-arrow-down" />
+            </Button>
+          </DropdownToggle>
+
+          <Dropdown.Menu>
+            {boards.map(board => {
+              if (board._id !== currentBoard._id) {
```
Comment (batamar) [inline src/modules/deals/components/Board.js]: write renderBoard function

File: src/modules/deals/components/Pipeline.js [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+          droppableId={pipeline._id}
+        >
+          {provided => (
+            <PipelineBody
+              innerRef={provided.innerRef}
+              {...provided.droppableProps}
+            >
+              <div>
+                {stages.map((stage, index) => {
+                  return (
```
Comment (batamar) [inline src/modules/deals/components/Pipeline.js]: write seperate renderStage function

File: src/modules/deals/components/Stage.js [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+
+  showForm() {
+    this.setState({ show: true });
+  }
+
+  closeForm() {
+    this.setState({ show: false });
+  }
+
+  renderAmount(amount) {
```
Comment (batamar) [inline src/modules/deals/components/Stage.js]: refactor this function

File: src/modules/deals/components/Stage.js [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+                      {__('Deal')}: {deals.length}
+                    </span>
+                  </div>
+
+                  {this.renderAmount(stage.amount)}
+                </StageHeader>
+
+                <StageBody>
+                  <Droppable droppableId={stage._id} type="stage">
+                    {dropProvided => (
```
Comment (batamar) [inline src/modules/deals/components/Stage.js]: write separate function

File: src/modules/deals/containers/Pipeline.js [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+    // if move to other pipeline, will change pipelineId
+    const stagesChange = (_id, pipelineId) => {
+      stagesChangeMutation({
+        variables: { _id, pipelineId }
+      });
+    };
+
+    const extendedProps = {
+      ...this.props,
+      stagesFromDb,
```
[the remark this answers is not in the document]
Comment (munkhjin0223) [reply src/modules/deals/containers/Pipeline.js]: No. But used above

File: src/modules/settings/deals/components/PipelineForm.js [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+          <ControlLabel>Name</ControlLabel>
+
+          <FormControl
+            id="pipeline-name"
+            defaultValue={object.name}
+            type="text"
+            required
+          />
+        </FormGroup>
+        {pipeline ? (
```
Comment (batamar) [inline src/modules/settings/deals/components/PipelineForm.js]: write renderStages function

File: src/modules/deals/containers/Stage.js [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+    if (this.props.state !== nextProps.state) {
+      const {
+        state: { type, index, itemId },
+        stageId,
+        dealsUpdateOrder,
+        dealsChange,
+        stageDetailQuery,
+        dealsQuery
+      } = nextProps;
+      const { deals } = this.state;
```
Comment (batamar) [inline src/modules/deals/containers/Stage.js]: Add line above

File: src/modules/deals/containers/Stage.js [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+
+  // create or update deal
+  saveDeal(doc, callback, deal) {
+    const {
+      addMutation,
+      editMutation,
+      stageDetailQuery,
+      dealsQuery
+    } = this.props;
+    const { deals } = this.state;
```
Comment (batamar) [inline src/modules/deals/containers/Stage.js]: Add line above

File: src/modules/deals/containers/Pipeline.js [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.state !== nextProps.state) {
+      const {
+        state: { type, index, itemId },
+        pipeline,
+        stagesUpdateOrder,
+        stagesChange
+      } = nextProps;
+      const { stages } = this.state;
```
Comment (batamar) [inline src/modules/deals/containers/Pipeline.js]: Add line above

File: src/modules/deals/components/common/ItemCounter.js [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+
+const propTypes = {
+  items: PropTypes.array
+};
+
+class ItemCounter extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
```
Comment (batamar) [inline src/modules/deals/components/common/ItemCounter.js]: Write in 1 line

File: src/modules/deals/components/common/ItemCounter.js [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+      show: false
+    };
+  }
+
+  showOthers() {
+    this.setState({
```
Comment (batamar) [inline src/modules/deals/components/common/ItemCounter.js]: Write in 1 line

File: src/modules/deals/components/deal/DealMoveForm.js [hunk 33]
```diff
@@ -0,0 +1,138 @@
...
+import { selectOptions } from '../../utils';
+import { DealMoveFormContainer } from '../../styles';
+
+class DealMoveForm extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.move = this.move.bind(this);
+
+    this.state = {
```
Comment (batamar) [inline src/modules/deals/components/deal/DealMoveForm.js]: Write in 1 line

File: src/modules/deals/components/deal/DealMoveForm.js [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+
+  render() {
+    const { __ } = this.context;
+    const { boards, pipelines, stages, boardId, pipelineId } = this.props;
+
+    return (
+      <DealMoveFormContainer>
+        <form>
+          <FormGroup>
+            <Select
```
Comment (batamar) [inline src/modules/deals/components/deal/DealMoveForm.js]: Write common function for this repeated Select. optionRenderer

File: src/modules/deals/containers/Pipeline.js [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+    if (this.props.state !== nextProps.state) {
+      const { state: { type, index, itemId }, pipeline } = nextProps;
+
+      const { stages } = this.state;
+
+      if (type === 'removeItem') {
+        // Remove from list
+        stages.splice(index, 1);
+
+        this.stagesUpdateOrder(collectOrders(stages));
```
Comment (batamar) [inline src/modules/deals/containers/Pipeline.js]: Remove this function and run function's body  once above setState

File: src/modules/deals/containers/Pipeline.js [hunk 36]
```diff
@@ -0,0 +1,131 @@
...
+
+    stagesUpdateOrderMutation({
+      variables: { orders }
+    }).catch(error => {
+      Alert.error(error.message);
+    });
+  }
+
+  // if move to other pipeline, will change pipelineId
+  stagesChange(_id, pipelineId) {
```
Comment (batamar) [inline src/modules/deals/containers/Pipeline.js]: Do we really this function ?. Just run it's body in type === 'addItem' ?

File: src/modules/deals/containers/Stage.js [hunk 37]
```diff
@@ -0,0 +1,278 @@
...
+        dealsQuery
+      } = nextProps;
+
+      const { deals } = this.state;
+
+      if (type === 'removeItem') {
+        // Remove from list
+        deals.splice(index, 1);
+
+        this.dealsUpdateOrder(collectOrders(deals), () => {
```
Comment (batamar) [inline src/modules/deals/containers/Stage.js]: Remove this function and run function's body once above setState

File: src/modules/deals/containers/Stage.js [hunk 38]
```diff
@@ -0,0 +1,278 @@
...
+  }),
+  graphql(gql(mutations.dealsUpdateOrder), {
+    name: 'dealsUpdateOrderMutation'
+  }),
+  graphql(gql(mutations.dealsChange), {
+    name: 'dealsChangeMutation'
+  })
+)(StageContainer);
+
+class StageWithDeals extends React.Component {
```
Comment (batamar) [inline src/modules/deals/containers/Stage.js]: Do we need this container ?

**Comment by batamar** [comment]
bug


When this is no boards

**Pull request merged**